### PR TITLE
fix: validation error when ! is in the path

### LINKED ISSUE #5320: Unclear error message in case of Exclamation mark in project path breaking the build on Windows
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
On Windows build if the path of the project folder contains "!", webpack will throw error:
```
    ERROR in Entry module not found: Error: Can't resolve 'C:\' in 'C:\!ParentFolder\MyProjectFolder'
```
This bug was already reported #4906 with steps to reproduce. And it was closed, I'm opening this one to attract attention that this issue creates hard time for new users of webpack who do not know about this limitation, since webpack does not give any clue to the user about it. It took an hour of my time to figure out why it's not working.

**If the current behavior is a bug, please provide the steps to reproduce.**
see #4906
or:
1) Take any project with webpack (e.g. https://github.com/pmunin/handsontable)
2) Put (or clone) it in any folder, with the name starting from "!" (exclamation mark)
3) run webpack build (e.g. "npm run build:umd" for repo provided at step 1)

**What is the expected behavior?**
Webpack should at least provide a proper error message saying that user cannot have Exclamation marks in the path.

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**
OS: Windows 10 x64
node v6.3.1
npm v3.10.3
webpack v2.6.1

### LINKED ISSUE #6742: webpack generates error when run in a directory starting with an exclamation sign
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
Using webpack in a directory start starts with `!`, generates errors but works as expected when the directory does not start with `!`

```

D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test>npm run build

> example@1.0.0 build D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test
> webpack --progress --display-error-details

Hash: 03eec963e3df63b69ec7
Version: webpack 3.11.0
Time: 3824ms
   Asset     Size  Chunks             Chunk Names
index.js  7.37 kB       0  [emitted]  index
   [0] ./src/index.js 21 bytes {0} [built]
   [1] ./src/index.css 4.48 kB {0} [built] [failed] [1 error]

ERROR in ./src/index.css
Module build failed: EntryModuleNotFoundError: Entry module not found: Error: Can't resolve 'test\src\index.css' in 'D:\MyDev\ljs_app\trunk\periscope\examples\w
ebpack\!test\src'
    at moduleFactory.create (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\webpack\lib\Compilation.js:411:30)
    at factory (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\webpack\lib\NormalModuleFactory.js:237:20)
    at resolver (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\webpack\lib\NormalModuleFactory.js:60:20)
    at asyncLib.parallel (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\webpack\lib\NormalModuleFactory.js:127:20)
    at D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\async\dist\async.js:3874:9
    at D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\async\dist\async.js:473:16
    at iteratorCallback (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\async\dist\async.js:1048:13)
    at D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\async\dist\async.js:958:16
    at D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\async\dist\async.js:3871:13
    at resolvers.normal.resolve (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\webpack\lib\NormalModuleFactory.js:119:22)
    at onError (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:65:10)
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at runAfter (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:158:4)
    at innerCallback (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:146:3)
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at next (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\tapable\lib\Tapable.js:252:11)
    at D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\UnsafeCachePlugin.js:40:4
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at runAfter (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:158:4)
    at innerCallback (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:146:3)
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at next (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\tapable\lib\Tapable.js:252:11)
    at innerCallback (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:144:11)
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at next (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\tapable\lib\Tapable.js:249:35)
    at resolver.doResolve.createInnerCallback (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\DescriptionFilePlugin.j
s:44:6)
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at afterInnerCallback (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\Resolver.js:166:11)
    at loggingCallbackWrapper (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\enhanced-resolve\lib\createInnerCallback.js:31:19)
    at next (D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\tapable\lib\Tapable.js:249:35)
 @ ./src/index.css
 @ ./src/index.js

ERROR in ./src/index.css
Module not found: Error: Can't resolve 'test/node_modules/css-loader/lib/css-base.js' in 'D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src'
resolve 'test/node_modules/css-loader/lib/css-base.js' in 'D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src'
  Parsed request is a module
  using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./src)
    Field 'browser' doesn't contain a valid alias configuration
  after using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./src)
    resolve as module
      D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src\node_modules doesn't exist or is not a directory
      D:\MyDev\ljs_app\trunk\periscope\examples\webpack\node_modules doesn't exist or is not a directory
      D:\MyDev\ljs_app\trunk\periscope\examples\node_modules doesn't exist or is not a directory
      D:\MyDev\ljs_app\trunk\periscope\node_modules doesn't exist or is not a directory
      D:\MyDev\ljs_app\trunk\node_modules doesn't exist or is not a directory
      D:\MyDev\ljs_app\node_modules doesn't exist or is not a directory
      D:\MyDev\node_modules doesn't exist or is not a directory
      D:\node_modules doesn't exist or is not a directory
      looking for modules in D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules
        using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./node_modules)
          Field 'browser' doesn't contain a valid alias configuration
        after using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./node_modules)
          using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./node_modules/test/node_modules/css-load
er/lib/css-base.js)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js.json doesn't exist
            as directory
              D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js doesn't exist
[D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src\node_modules]
[D:\MyDev\ljs_app\trunk\periscope\examples\webpack\node_modules]
[D:\MyDev\ljs_app\trunk\periscope\examples\node_modules]
[D:\MyDev\ljs_app\trunk\periscope\node_modules]
[D:\MyDev\ljs_app\trunk\node_modules]
[D:\MyDev\ljs_app\node_modules]
[D:\MyDev\node_modules]
[D:\node_modules]
[D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js]
[D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js.js]
[D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js.json]
[D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\node_modules\css-loader\lib\css-base.js]
 @ ./src/index.css 1:27-88
 @ ./src/index.js
Child extract-text-webpack-plugin ..!test\node_modules\extract-text-webpack-plugin\dist ..!test\node_modules\css-loader\index.js??ref--1-1!..!test\src\index.css
:

    ERROR in Entry module not found: Error: Can't resolve 'test\src\index.css' in 'D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src'
    resolve 'test\src\index.css' in 'D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src'
      Parsed request is a module
      using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./src)
        Field 'browser' doesn't contain a valid alias configuration
      after using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./src)
        resolve as module
          D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\src\node_modules doesn't exist or is not a directory
          D:\MyDev\ljs_app\trunk\periscope\examples\webpack\node_modules doesn't exist or is not a directory
          D:\MyDev\ljs_app\trunk\periscope\examples\node_modules doesn't exist or is not a directory
          D:\MyDev\ljs_app\trunk\periscope\node_modules doesn't exist or is not a directory
          D:\MyDev\ljs_app\trunk\node_modules doesn't exist or is not a directory
          D:\MyDev\ljs_app\node_modules doesn't exist or is not a directory
          D:\MyDev\node_modules doesn't exist or is not a directory
          D:\node_modules doesn't exist or is not a directory
          looking for modules in D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules
            using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./node_modules)
              Field 'browser' doesn't contain a valid alias configuration
            after using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./node_modules)
              using description file: D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\package.json (relative path: ./node_modules/test/src/index.css)
                no extension
                  Field 'browser' doesn't contain a valid alias configuration
                  D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\src\index.css doesn't exist
                .js
                  Field 'browser' doesn't contain a valid alias configuration
                  D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\src\index.css.js doesn't exist
                .json
                  Field 'browser' doesn't contain a valid alias configuration
                  D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\src\index.css.json doesn't exist
                as directory
                  D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test\node_modules\test\src\index.css doesn't exist
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! example@1.0.0 build: `webpack --progress --display-error-details`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the example@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\doberkofler.LBITS\AppData\Roaming\npm-cache\_logs\2018-03-13T16_50_39_385Z-debug.log

D:\MyDev\ljs_app\trunk\periscope\examples\webpack\!test>
```

**If the current behavior is a bug, please provide the steps to reproduce.**
just run `webpack --progress --display-error-details` in a directory that starts with an exclamation sign.

**minimal reproduceable repo:**
https://github.com/doberkofler/webpack_issue_6742

webpack.config.js:
```
'use strict';

const path = require('path');
const webpack = require('webpack');
const ExtractTextPlugin = require('extract-text-webpack-plugin');

module.exports = function config(env) {
	return {
		entry: {
			index: './src/index.js'
		},
		output: {
			path: path.resolve(__dirname, 'dist'),
			filename: '[name].js'
		},
		module: {
			rules: [
				{
					test: /\.js$/i,
					include: /src/,
					exclude: /node_modules/,
					use: {
						loader: 'babel-loader'
					}
				},
				{
					test: /\.css$/i,
					include: /src/,
					exclude: /node_modules/,
					use: ExtractTextPlugin.extract([{
						loader: 'css-loader',
						options: {
							sourceMap: false
						}
					}])
				}
			]
		},
		plugins: [
			new ExtractTextPlugin('[name].css')
		]
	};
};
```

**What is the expected behavior?**
webpack should not fail or at least report that it cannot be used in certain directories.

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**
node: v8.10.0
webpack: 3.11.0

### PR DESCRIPTION
**Pull request by byzyk**
**What kind of change does this PR introduce?**

bugfix

**Did you add tests for your changes?**

yes

**If relevant, link to documentation update:**

- [ ] TODO: possibly add a warning here https://webpack.js.org/configuration/output/#output-path

**Summary**

Throw validation error when `!` was found in a path.
Closes #6742, closes #5320. 

**Does this PR introduce a breaking change?**

no

**Other information**

### CODE REVIEW

**Comment by byzyk** [comment]
Hi @sokra, I fixed regexp inside absolute path validator to be able to check if there is `!` present in a path. 

However, I can't test my changes because it seems no matter what I do in `schemas/ajv.absolutePath.js` it doesn't reflect. For example, I tried to change the error message but it had no effect. It's not the package linking problem I checked that a couple of times already.

Please advice are there any extra steps required to make this work or am I missing something?

**Comment by chengluyu** [comment]
I think you should add a new message when ! is included in absolute path. Here's my [solution](https://github.com/chengluyu/webpack/commit/c4255596318a698cd3253d23852996f5bcbcbaf9).

p.s. I wonder why people don't use `path.isAbsolute`.

File: schemas/ajv.absolutePath.js [hunk 1]
```diff
@@ -1,29 +1,33 @@
...
-module.exports = (ajv) => ajv.addKeyword("absolutePath", {
-	errors: true,
-	type: "string",
-	compile(expected, schema) {
-		function callback(data) {
-			const passes = expected === /^(?:[A-Za-z]:\\|\/)/.test(data);
-			if(!passes) {
-				callback.errors = [getErrorFor(expected, data, schema)];
+
+module.exports = ajv =>
```
Comment (montogeek) [inline schemas/ajv.absolutePath.js]: Why code format changed?

Comment (evilebottnawi) [reply schemas/ajv.absolutePath.js]: @montogeek `prettier`

Comment (montogeek) [reply schemas/ajv.absolutePath.js]: I understand that src code was already formatted with prettier, he didn't changed anything to increment line width

Comment (ooflorent) [reply schemas/ajv.absolutePath.js]: This file never was reformatted by prettier.

Comment (sokra) [reply schemas/ajv.absolutePath.js]: looks like the `schemas` folder is not included in the prettier scripts.

Comment (byzyk) [reply schemas/ajv.absolutePath.js]: Yes, the file was never formatted, when I first opened it there were 19 ESLint violations.

**Comment by byzyk** [comment]
hey @sokra, I resolved some conflicts, does it look ok to be merged now?

Also, what would be the right place for docs update? Should we add a warning in all related places (eg. `output.path`, `output.filename`)?

**Comment by sokra** [comment]
Thanks

**Comment by sokra** [comment]
Thanks

**Pull request merged**